Add predictable ID for help fields, aria describedby attributes

ArnoSaine opened the pull request:
### Reasons for making this change

Make screen reader read field errors, description and help text.

Fixes #959

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### Deploy preview

Once CI finishes, you should be able to view a deploy preview of the playground from this PR at this link: https://deploy-preview-[PR_NUMBER]--rjsf.netlify.app/

epicfaace commented:
Thanks for the PR. I wonder if we can merge the changes done in https://github.com/rjsf-team/react-jsonschema-form/pull/2360 into this. Specifically, can you make sure that all the help / description / title field ids are using the id utility functions you defined in utils.js? Specifically, we might need to change `packages/semantic-ui/src/FieldTemplate/FieldTemplate.js`, for example.

Can we also add snapshot tests as needed for the affected themes, because now the help fields have different ids?

ArnoSaine commented:
Thank you for the suggestions. I searched for other mentions of `__description`, `__help` and `__errors`. `packages/semantic-ui/src/FieldTemplate/FieldTemplate.js` was the only one not using the utility function. it's now fixed.

I updated the affected snapshots and added more material-ui updates.

epicfaace commented on packages/core/src/utils.js:
```diff
@@ -1265,3 +1265,13 @@ export function schemaRequiresTrueValue(schema) {
+export const errorsId = id => `${id}__errors`;
+
+export const ariaDescribedBy = id => ({
+  "aria-describedby": `${errorsId(id)} ${descriptionId(id)} ${helpId(id)}`,
```
What if, for example, the help text is empty? Would it be correct in that case to still include helpId within aria-describedby, or should it be removed in that case?

ArnoSaine replied:
It is an issue. As far as I know, the specification does not say anything about missing element references. Still, https://validator.w3.org/ gives an error, if for example help element is missing.

VoiceOver reads only the existing elements without a problem when the input gets focus, as expected. I guess other tools work similarly. Order of the IDs also matter: I think it's best to have errors first. If field has errors, user gets that infomation first, without need to go through description and help every time.

I'd suggest we add all IDs, because it still improves accessibility.

epicfaace replied:
I see, thanks!

epicfaace commented on packages/core/src/utils.js:
```diff
@@ -1265,3 +1265,13 @@ export function schemaRequiresTrueValue(schema) {
+export const errorsId = id => `${id}__errors`;
+
+export const ariaDescribedBy = id => ({
+  "aria-describedby": `${errorsId(id)} ${descriptionId(id)} ${helpId(id)}`,
```
For consistency, I wonder if it's better to just return a string here instead of a dict.

ArnoSaine replied:
Done! You are right. It's better that way.

epicfaace commented on packages/core/src/components/fields/SchemaField.js:
```diff
@@ -68,10 +68,7 @@ function Label(props) {
-      className="control-label"
-      htmlFor={id}
-      aria-describedby={ariaDescribedBy(id)}>
+    <label className="control-label" htmlFor={id}>
```
Why was the original aria-describedby here incorrect?

ArnoSaine replied:
I added it by mistake here: https://github.com/rjsf-team/react-jsonschema-form/pull/2386/commits/d332de122ed74dd534334d0c2835a23ce1577811

Original is correct: `<input>` has `id`, `<label>` has `for`. `aria-describedby` goes also to the `<input>`

epicfaace commented on packages/material-ui/src/FieldTemplate/FieldTemplate.tsx:
```diff
@@ -43,22 +43,27 @@ const FieldTemplate = ({
               return (
                 <ListItem key={i} disableGutters={true}>
-                  <FormHelperText id={id}>{error}</FormHelperText>
+                  <FormHelperText>{error}</FormHelperText>
```
Why remove the id here?

ArnoSaine replied:
It was inside a `.map`. If there were multiple errors, it was not unique. Same id was also used with the `rawHelp` text. I moved the id to the parent `<List>` element and added suffix to both help and error list ids.

epicfaace replied:
Thanks! Can you add a short section in the "v3 upgrade guide" doc (within the documentation) that describes some of these breaking changes in id's? Just a quick list of what exactly were changed, so that someone who upgrades to v3 will know how the ids will change in the form.

ArnoSaine replied:
Could we consider this id change to be an internal fix? Element ids must be unique within the document and it is an undocumented feature. Here's a screenshot of the issue from the playground:
<img width="677" alt="Screenshot 2021-05-24 at 15 50 29" src="https://user-images.githubusercontent.com/5928960/119351770-88781c80-bca9-11eb-9da0-bad2730ee06d.png">

In this PR we let the input use the generated `id` (no change). To fix the case of non-unique ids and make `aria-describedby` behave correctly, we suffix help and errors ids (in material-ui FieldTemplate).

epicfaace commented on packages/core/index.d.ts:
```diff
@@ -433,6 +433,14 @@ declare module '@rjsf/core' {
 
         export function schemaRequiresTrueValue(schema: JSONSchema7): boolean;
+
+        export function descriptionId(id: string): string;
```
Let's add one-line comments before each new function explaining what they do.

ArnoSaine replied:
Sure! I commented `packages/core/src/utils.js` as most of the comments are in that file.

epicfaace commented:
Hi @ArnoSaine it's been a while, do you remember where we are with this PR? Is this ready except for making it a breaking change / adding it to a v5.x upgrade guide?

heath-freenome commented:
Reimplemented in #3380
